### DISCUSSION TITLE
Should we use Typescript

### DISCUSSION THREAD

**Discussion by zhuiks** [ezra-bible-app/ezra-bible-app#264] [Development]
I just want to open up a discussion about [typescript](https://www.typescriptlang.org/). What a pros and cons.

The biggest con is the cost of refactoring. But I guess it can be a gradual transition. Or no transition at all. We can use typescript just for the new components

#### The WHY
Static typing results in better code quality and development experience.

**Comment by tobias-klein** [comment]
I think another con is that there are probably still more developers familiar just with JavaScript and not Typescript. I feel like with JavaScript the entry barrier is lower and the eco system is bigger. I would suggest to give this discussion some more time based on your own experience developing/maintaining Ezra components. The question is also whether we would avoid certain bugs in the course of time. I am familiar both with statically typed languages (C++, Java) as well as dynamically typed languages (Python, JavaScript) and I have to say that I personally prefer the dynamic typing over static typing for flexibility reasons. But I think the even bigger argument is the ecosystem / entry barrier.

**Comment by zhuiks** [comment]
I realized that the most thing I want from typescript is `@type` annotations. They are very helpful for navigating the code and understanding function expectations. But should not be the typescript. It could be just JSDoc comments. I thing most of the code editors including VSCode will pick up on those.

**Reply by tobias-klein** [reply]
We can make it a pattern to be more consistent about adding JSDoc comments when a function's purpose/interface is not self explaining 🙂.

**Reply by zhuiks** [reply]
Sometimes it's just nice to use that click/hover over function name VSCode feature to get to the source of the function. And without type annotations VSCode can't track the source. I use global search by function name a lot, but it's tiresome.
